### PR TITLE
[FIX] records: allow change of comodel for comp dep fields

### PR DESCRIPTION
**Pull request by jjmaksoud**
When replacing references by batch, the assertion that the model source == model dest is unique for company dependent fields, other indirect references get their model updated in this case.
So we can remove the assertion and change the comodel for jsonb company dependent fields too.

tbg-1590

### CODE REVIEW

**Comment by jjmaksoud** [comment]
upgradeci retry with always only documents_project_sale in version 18.0

File: src/util/records.py [hunk 1]
```diff
@@ -1479,7 +1479,15 @@ def replace_record_references_batch(cr, id_mapping, model_src, model_dst=None, r
             continue
         if ir.company_dependent_comodel:
             if ir.company_dependent_comodel == model_src:
-                assert model_src == model_dst
+                if model_src != model_dst:
```
Comment (aj-fuentes) [inline src/util/records.py:1482]: Can't we just remove the `assert`? We can check instead that _all_ keys were replaced. In such cases we can "trust" that the model of the underlying field is being changed. I find it odd to update ir_model_fields relation column here...
@KangOl wdyt?

Comment (jjmaksoud) [reply src/util/records.py:1482]: we can remove the assertion and trust that the relation will change later, I only added this part to be consistent with the other IR where we change the res_model, and I share the same concern

Comment (aj-fuentes) [reply src/util/records.py:1482]: > I only added this part to be consistent with the other IR where we change the res_model

Can you put a link to it please?

Comment (jjmaksoud) [reply src/util/records.py:1482]: https://github.com/odoo/upgrade-util/blob/c47b904633f1e3133c277b339ffeb534bffa056d/src/util/records.py#L1508-L1531

Comment (aj-fuentes) [reply src/util/records.py:1482]: Well indeed. To be aligned we could assume that the "ir.res_model" of company dependent fields (jsonb version) is the model of the field. I think we could still add some sort of check after we change the model to ensure all values (ids) are present in the referenced table. Can you do a quick test to see in 18 what happens if the value points to an invalid key (id)? Is it ignored? Does it raises any error?

File: src/util/records.py [hunk 2]
```diff
@@ -1479,7 +1479,15 @@ def replace_record_references_batch(cr, id_mapping, model_src, model_dst=None, r
...
+                    cr.execute(
+                        "UPDATE ir_model_fields SET relation = %s WHERE model = %s AND name = %s",
+                        (
+                            model_dst,
+                            model_of_table(cr, ir.table),
+                            ir.res_id,
+                        ),
```
Comment (aj-fuentes) [inline src/util/records.py:1489]: nit: IMHO it helps readability by reducing chances of confusing with a function call. Feel free to ignore it.
```suggestion
                        [
                            model_dst,
                            model_of_table(cr, ir.table),
                            ir.res_id,
                        ],
```

File: src/util/records.py (hunk 1 above)
Comment (jjmaksoud) [reply src/util/records.py:1482]: Did a small test, if the id does not exist then an [error](https://github.com/odoo/odoo/blob/3e4c4516f7d238e76e2e9b7b71a807f7cc7d55b3/odoo/fields.py#L1267-L1271) is raised and the view is inaccessible, i'll add the check, and I will do the same test for regular IR fields

File: src/util/records.py [hunk 3]
```diff
@@ -1504,6 +1512,30 @@ def replace_record_references_batch(cr, id_mapping, model_src, model_dst=None, r
...
+                        """,
+                        table=ir.table,
+                        column=ir.res_id,
+                        dest_table=table_of_model(cr, model_dst),
+                    )
+                )
+                invalid_ref = list(cr.fetchall())
+                assert not invalid_ref, "Invalid company dependent values for {}.{} referencing model {}: {}".format(
+                    model_of_table(cr, ir.table), ir.res_id, model_dst, invalid_ref
+                )
```
Comment (KangOl) [inline src/util/records.py:1538]:
```suggestion
                if invalid_ref:
                    raise RuntimeError("Invalid company dependent values for {}.{} referencing model {}: {}".format(model_of_table(cr, ir.table), ir.res_id, model_dst, invalid_ref))
```

NOTE: the suggestion will still need to be formatted correctly.

**Review by KangOl — APPROVED** [review]
@robodoo r+

File: src/util/records.py [hunk 4]
```diff
@@ -1504,6 +1512,33 @@ def replace_record_references_batch(cr, id_mapping, model_src, model_dst=None, r
                     [list(map(list, id_mapping.items()))],
                 ).decode()
                 explode_execute(cr, query, table=ir.table)
+                # ensure all new ids exist
+                cr.execute(
+                    format_query(
+                        cr,
+                        """
+                        Select t.id AS id,
```
Comment (aj-fuentes) [inline src/util/records.py:1520]: 🙀 
```suggestion
                        SELECT t.id AS id,
```

Comment (jjmaksoud) [reply src/util/records.py:1520]: :scream_cat:

**Pull request closed without merging**